packages/apps/desktop-viewer-test/src/frontend/components/modelSelector/SelectIModel.tsx

```diff
@@ -5,34 +5,194 @@
 
 import "./SelectIModel.scss";
 
+import { BriefcaseConnection } from "@bentley/imodeljs-frontend";
+import { getBriefcaseStatus, ModelStatus } from "@itwin/desktop-viewer-react";
 import {
   IModelFull,
   IModelGrid,
   IModelGridProps,
 } from "@itwin/imodel-browser-react";
-import { Title } from "@itwin/itwinui-react";
+import { TileProps, Title } from "@itwin/itwinui-react";
 import { useNavigate } from "@reach/router";
-import React, { useContext } from "react";
+import React, {
+  useCallback,
+  useContext,
+  useEffect,
+  useMemo,
+  useState,
+} from "react";
 
+import { useDownload } from "../../hooks/useDownload";
 import { SettingsContext } from "../../services/SettingsClient";
+import { IModelContext } from "../routes";
+import { BriefcaseStatus } from "./BriefcaseStatus";
 
 interface SelectIModelProps extends IModelGridProps {
   projectName?: string;
 }
 
+const useProgressIndicator = (iModel: IModelFull) => {
+  const userSettings = useContext(SettingsContext);
+  const [status, setStatus] = useState<ModelStatus>();
+  const [briefcase, setBriefcase] = useState<BriefcaseConnection>();
+  const modelContext = useContext(IModelContext);
+  const navigate = useNavigate();
+
+  /**
+   * Get the local file from settings
+   * @returns
+   */
+  const getLocal = useCallback(() => {
+    const recents = userSettings.settings.recents;
+    if (recents) {
+      return recents.find((recent) => {
+        return (
+          recent.iTwinId === iModel.projectId && recent.iModelId === iModel.id
+        );
+      });
+    }
+  }, [userSettings]);
+
+  const getBriefcase = useCallback(async () => {
+    // if there is a local file, open a briefcase connection and store it in state
+    const local = getLocal();
+    if (local?.path) {
+      const connection = await BriefcaseConnection.openFile({
+        fileName: local.path,
+        readonly: true,
+      });
+      setBriefcase(connection);
+    } else {
+      setStatus(ModelStatus.ONLINE);
+    }
+  }, []);
+
+  const { progress, doDownload } = useDownload(
+    iModel.id,
+    iModel.name ?? iModel.id,
+    iModel.projectId ?? ""
+  );
+
+  const getLatestChangesets = useCallback(async () => {
+    if (briefcase) {
+      await briefcase.pullAndMergeChanges();
+    }
+  }, [briefcase]);
+
+  const startDownload = useCallback(async () => {
+    try {
+      setStatus(ModelStatus.DOWNLOADING);
+      const fileName = await doDownload();
+      if (fileName) {
+        setStatus(ModelStatus.UPTODATE);
+      } else {
+        setStatus(ModelStatus.ONLINE);
+      }
+      return fileName;
+    } catch {
+      setStatus(ModelStatus.ERROR);
+    }
+  }, []);
+
+  const mergeChanges = useCallback(() => {
+    setStatus(ModelStatus.MERGING);
```

Comment:
same comments as the status bar item, couldnt this call getLatestChangesets?

Reply:
it could be since we have UI tied to the status state, this ensures that updates occur in the proper timeframe. For example, if we set the status to merging, and then call getLatestChangesets, and then change the status again immediately after, the UI would not update appropriately (or timely I should say). This ensures that getLatestChangesets executes once the spinner is rendered